File: backend/RAG_python-quiz/app/routers/quiz.py
from typing import List, Optional

from fastapi import APIRouter, Body, Depends, Form, Query

from app.logger import get_logger
from app.routers import quiz_helpers
from app.routers.service_helpers import run_async_service
from app.services.pg import pg_access_control as pg_service
from app.services.pg import pg_quiz_service
from app.services.pg.pg_classes_service import is_user_teacher
from app.services.ai.quiz_feedback_service import generate_quiz_feedback_text
from app.services.assessment.quiz_generation_service import QuizGenerateResponse, generate_quiz_from_files
from app.utils.aqg import (
    DIFFICULTY_TO_BLOOM,
    BloomLevel,
    Difficulty,
)
from app.utils.jwt_utils import get_current_user

logger = get_logger(__name__)
router = APIRouter(prefix="", tags=["quiz"])


def _questions_from_payload(payload: dict, *, required: bool) -> list:
    return quiz_helpers.questions_from_payload(payload, required=required)


async def _run_quiz_service(func, *args, action: str, fallback_detail=None):
    return await quiz_helpers.run_quiz_service(
        func,
        *args,
        action=action,
        logger=logger,
        fallback_detail=fallback_detail,
    )


@router.post("/generate", response_model=QuizGenerateResponse)
async def generate_quiz(
    file_ids: List[str] = Form(...),
    bloom_levels: Optional[List[BloomLevel]] = Form(None),
    difficulty: Optional[Difficulty] = Form(None),
    num_questions: int = Form(5),
    user: dict = Depends(get_current_user),
):
    quiz_helpers.require_quiz_teacher(user, "Only teachers can generate quizzes", is_user_teacher)
    quiz_helpers.require_document_access(user, file_ids, pg_service.can_manage_documents)
    return await generate_quiz_from_files(file_ids, bloom_levels, difficulty, num_questions)


@router.get("/bloom-levels")
async def get_bloom_levels():
    from app.utils.aqg import BLOOM_DESCRIPTIONS

    return {
        "levels": [
            {"value": level, "description": description}
            for level, description in BLOOM_DESCRIPTIONS.items()
        ]
    }


@router.get("/difficulties")
async def get_difficulties():
    return {
        "difficulties": [
            {"value": difficulty, "bloom_levels": levels}
            for difficulty, levels in DIFFICULTY_TO_BLOOM.items()
        ]
    }


@router.get("/list")
async def get_all_quizzes(
    class_id: str = Query(...),
    user: dict = Depends(get_current_user),
):
    quiz_helpers.require_class_access(user, class_id, pg_service.can_access_class)
    quizzes = await _run_quiz_service(
        pg_quiz_service.get_quizzes_by_class,
        class_id,
        user["user_id"],
        action="Get quizzes",
    )
    return {"message": "Fetched quizzes", "quizzes": quizzes, "total": len(quizzes)}


@router.get("/{quiz_id}")
async def get_quiz(
    quiz_id: str,
    user: dict = Depends(get_current_user),
):
    quiz_payload = await _run_quiz_service(
        pg_quiz_service.get_quiz_by_id,
        quiz_id,
        user["user_id"],
        action="Get quiz",
    )
    return {"message": "Fetched quiz", "quiz": quiz_payload}


@router.post("/")
async def create_quiz(
    payload: dict = Body(...),
    user: dict = Depends(get_current_user),
):
    quiz_helpers.require_quiz_teacher(user, "Only teachers can create quizzes", is_user_teacher)
    questions = _questions_from_payload(payload, required=False)
    file_ids = payload.get("file_ids") or []
    class_id = payload.get("class_id")
    quiz_helpers.require_create_source_access(
        user,
        class_id=class_id,
        file_ids=file_ids,
        class_checker=pg_service.can_access_class,
        document_checker=pg_service.can_manage_documents,
    )
    saved = await _run_quiz_service(
        pg_quiz_service.save_quiz,
        {"questions": questions},
        file_ids,
        payload.get("name"),
        class_id,
        action="Create quiz",
    )
    return {"message": "Quiz created", "quiz": saved}


@router.put("/{quiz_id}")
async def update_quiz(
    quiz_id: str,
    payload: dict = Body(...),
    user: dict = Depends(get_current_user),
):
    quiz_helpers.require_quiz_teacher(user, "Only teachers can update quizzes", is_user_teacher)
    quiz_helpers.require_quiz_access(user, quiz_id, pg_service.can_access_quiz)
    questions = _questions_from_payload(payload, required=True)
    if payload.get("file_ids"):
        quiz_helpers.require_document_access(user, payload["file_ids"], pg_service.can_manage_documents)
    updated = await _run_quiz_service(
        pg_quiz_service.update_quiz,
        quiz_id,
        {"questions": questions},
        payload.get("name"),
        payload.get("file_ids") if "file_ids" in payload else None,
        action="Update quiz",
    )
    return {"message": "Quiz updated", "quiz": updated}


@router.delete("/{quiz_id}")
async def delete_quiz(
    quiz_id: str,
    user: dict = Depends(get_current_user),
):
    quiz_helpers.require_quiz_teacher(user, "Only teachers can delete quizzes", is_user_teacher)
    quiz_helpers.require_quiz_access(user, quiz_id, pg_service.can_access_quiz)
    return await _run_quiz_service(
        pg_quiz_service.delete_quiz,
        quiz_id,
        user["user_id"],
        action="Delete quiz",
    )


@router.post("/{quiz_id}/submit")
async def submit_quiz(
    quiz_id: str,
    payload: dict = Body(...),
    user: dict = Depends(get_current_user),
):
    quiz_helpers.require_student(user, pg_service.is_user_student)
    quiz_helpers.require_quiz_access(user, quiz_id, pg_service.can_access_quiz)
    score, total_questions = quiz_helpers.require_score_fields(payload, "Score and total_questions required")

    return await _run_quiz_service(
        pg_quiz_service.submit_quiz_result,
        quiz_id,
        user["user_id"],
        payload.get("answers", []),
        score,
        total_questions,
        action="Submit quiz",
        fallback_detail=lambda error: str(error),
    )


@router.get("/{quiz_id}/results")
async def get_quiz_results(
    quiz_id: str,
    user: dict = Depends(get_current_user),
):
    quiz_helpers.require_quiz_teacher(user, "Only teachers can view all results", is_user_teacher)
    quiz_helpers.require_quiz_access(user, quiz_id, pg_service.can_access_quiz)
    results = await _run_quiz_service(
        pg_quiz_service.get_quiz_submissions,
        quiz_id,
        user["user_id"],
        action="Get quiz results",
        fallback_detail=lambda error: str(error),
    )
    return {"results": results}


@router.get("/{quiz_id}/my-result")
async def get_my_quiz_result(
    quiz_id: str,
    user: dict = Depends(get_current_user),
):
    quiz_helpers.require_quiz_access(user, quiz_id, pg_service.can_access_quiz)
    result = await _run_quiz_service(
        pg_quiz_service.get_student_quiz_submission,
        quiz_id,
        user["user_id"],
        action="Get my result",
        fallback_detail=lambda error: str(error),
    )
    return {"submission": result} if result else {"submission": None}


@router.post("/{quiz_id}/feedback")
async def generate_quiz_feedback(
    quiz_id: str,
    payload: dict = Body(...),
    user: dict = Depends(get_current_user),
):
    quiz_helpers.require_quiz_access(user, quiz_id, pg_service.can_access_quiz)
    score, total_questions = quiz_helpers.require_score_fields(payload, "score and total_questions are required")

    feedback_text = await run_async_service(
        generate_quiz_feedback_text,
        **quiz_helpers.feedback_kwargs(payload, score=score, total_questions=total_questions),
        logger=logger,
        log_message="Generate feedback failed: %s",
        fallback_detail=lambda error: str(error),
    )
    return {"feedback": feedback_text}

File: frontend/vite-project/src/api/quiz.js
import axios from 'axios';
import { API_BASE_URL } from '../config.js';
import { dedupe } from '../utils/requestDeduper';
import { getToken } from './auth';


/**
 * 生成測驗題目
 * @param {Array<string>} fileIds - Neo4j 文件 ID 列表
 * @param {Object} options - 可選參數
 * @param {Array<string>} options.bloomLevels - Bloom 認知層級列表
 * @param {string} options.difficulty - 難度級別 (easy/medium/difficult)
 * @param {number} options.numQuestions - 題目數量 (1-50)
 * @returns {Promise} - API 響應
 */
export const generateQuiz = async (fileIds, options = {}) => {
    const formData = new FormData();
    
    // 添加文件 IDs
    fileIds.forEach(id => {
        formData.append('file_ids', id);
    });
    
    // 添加可選參數
    if (options.bloomLevels && options.bloomLevels.length > 0) {
        options.bloomLevels.forEach(level => {
            formData.append('bloom_levels', level);
        });
    }
    
    if (options.difficulty) {
        formData.append('difficulty', options.difficulty);
    }
    
    if (options.numQuestions) {
        formData.append('num_questions', options.numQuestions);
    }
    
    console.log('發送測驗生成請求:', { fileIds, options });
    console.log('FormData entries:');
    for (let pair of formData.entries()) {
        console.log(pair[0] + ': ' + pair[1]);
    }
    
    return axios.post(`${API_BASE_URL}/quiz/generate`, formData, {
        headers: { 'Content-Type': 'multipart/form-data' }
    });
};


/**
 * 獲取所有 Bloom 認知層級
 * @returns {Promise} - Bloom 層級列表
 */
export const getBloomLevels = async () => {
    return axios.get(`${API_BASE_URL}/quiz/bloom-levels`);
};


/**
 * 獲取所有難度級別
 * @returns {Promise} - 難度級別列表
 */
export const getDifficulties = async () => {
    return axios.get(`${API_BASE_URL}/quiz/difficulties`);
};


/**
 * 獲取所有測驗列表
 * @param {string} classId - 班級 ID
 * @returns {Promise} - 測驗列表
 */
export const getAllQuizzes = async (classId, axiosConfig = {}) => {
    const key = `quiz:list:${classId || '__all__'}`;
    console.log(`[api.quiz] getAllQuizzes called, classId=${classId}, key=${key}`);
    // use dedupe utility; we keep a small TTL to merge very close sequential calls
    return dedupe(key, () => axios.get(`${API_BASE_URL}/quiz/list`, {
        params: { class_id: classId },
        ...axiosConfig,
    }), { ttl: 1000 });
};


/**
 * 根據 ID 獲取特定測驗
 * @param {string} quizId - 測驗 ID
 * @returns {Promise} - 測驗詳情
 */
export const getQuizById = async (quizId) => {
    return axios.get(`${API_BASE_URL}/quiz/${quizId}`);
};


/**
 * 刪除測驗
 * @param {string} quizId - 測驗 ID
 * @returns {Promise} - 刪除結果
 */
export const deleteQuiz = async (quizId) => {
    return axios.delete(`${API_BASE_URL}/quiz/${quizId}`);
};

/**
 * 建立新的測驗（教師手動建立或編輯後儲存）
 * @param {Object} quiz - 包含 name, questions (array), file_ids (optional), class_id (optional)
 */
export const createQuiz = async (quiz) => {
    return axios.post(`${API_BASE_URL}/quiz`, quiz);
};

/**
 * 更新現有測驗
 * @param {string} quizId
 * @param {Object} quiz - 同 createQuiz
 */
export const updateQuiz = async (quizId, quiz) => {
    return axios.put(`${API_BASE_URL}/quiz/${quizId}`, quiz);
};

/**
 * 提交測驗
 * @param {string} quizId
 * @param {Object} payload - { answers: [], score: number, total_questions: number }
 */
export const submitQuiz = async (quizId, payload) => {
    const token = getToken();
    const config = token ? { headers: { Authorization: `Bearer ${token}` } } : {};
    return axios.post(`${API_BASE_URL}/quiz/${quizId}/submit`, payload, config);
};

/**
 * 生成 AI 測驗回饋
 * @param {string} quizId
 * @param {Object} payload - { score, total_questions, percentage, bloom_summary, questions }
 */
export const generateQuizFeedback = async (quizId, payload) => {
    const token = getToken();
    const config = token ? { headers: { Authorization: `Bearer ${token}` } } : {};
    return axios.post(`${API_BASE_URL}/quiz/${quizId}/feedback`, payload, config);
};

/**
 * 獲取測驗結果（教師專用）
 * @param {string} quizId
 */
export const getQuizResults = async (quizId) => {
    const token = getToken();
    const config = token ? { headers: { Authorization: `Bearer ${token}` } } : {};
    return axios.get(`${API_BASE_URL}/quiz/${quizId}/results`, config);
};

/**
 * 獲取自己的測驗結果（學生專用）
 * @param {string} quizId
 */
export const getMyQuizResult = async (quizId) => {
    const token = getToken();
    const config = token ? { headers: { Authorization: `Bearer ${token}` } } : {};
    return axios.get(`${API_BASE_URL}/quiz/${quizId}/my-result`, config);
};
